src/dmd/dmodule.d

```diff
@@ -653,188 +653,199 @@ extern (C++) final class Module : Package
     // syntactic parse
     Module parse()
     {
-        //printf("Module::parse(srcfile='%s') this=%p\n", srcfile.name.toChars(), this);
-        const(char)* srcname = srcfile.name.toChars();
-        //printf("Module::parse(srcname = '%s')\n", srcname);
-        isPackageFile = (strcmp(srcfile.name.name(), "package.d") == 0 ||
-                         strcmp(srcfile.name.name(), "package.di") == 0);
-        char* buf = cast(char*)srcfile.buffer;
-        size_t buflen = srcfile.len;
-        if (buflen >= 2)
+
+
+        enum Endian { little, big}
+        enum SourceEncoding { utf16, utf32}
+
+        char[] toUtf8(SourceEncoding sourceEncoding, Endian endian)(const(char)[] buf)
```

Comment:
`toUtf8` (should be spelled `toUTF8`) tries to be generic, but there are only 4 possible instantiations. Inside the template, it uses `static if` everywhere to implement each of the 4 versions. It all seems rather a pointless abstraction. The thicket of `static if` makes it hard to read. Better to just have two separate functions.

Reply:
Also no Ddoc function comment.

Reply:
I pulled out the UTF16 code since it was the nested confusing part.  Do you still think it should be 2 functions? For future reference, do local functions generally need Ddoc comments?

Reply:
Local functions can use abbreviated ones.